### ISSUE TITLE
Add a no-RNG option and use different default RNG options based on difficulty for campaigns

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#7755]
The idea being to ease players into the randomness rather than having them immediately need to deal with things like a Mage missing all three attacks in a row. This isn't meant to mean that the different types of RNG actually make the campaign itself more difficult. Rather it's about the additional mental complexity in planning for failure - a lot of complaints seem to end up being because the player interpreted "high chance of success" with "100% chance of success" and then they aren't able to recover when their plan fails. So for example:
* Easy - default to no-RNG
* Normal - default to biased RNG
* Hard - default to full randomness (aka how things work currently)

The no-RNG option would just be a simple "damage multiplied by hit chance" calculation, not anything more complex.

This would apply only to campaigns, and each campaign difficulty would only need to be balanced against the default RNG option. If someone plays on Easy with full randomness and complains it's too hard, too bad. If someone plays on Hard with no-RNG and complains it's too easy, same answer.

**Comment by Dalas121** [comment]
I agree a no-RNG option is a good idea; the majority of complaints I've seen about Wesnoth center on RNG. I think it's a core and fun part of the game, but forcing it on everyone limits Wesnoth's audience.

I also agree it makes sense to balance around the default RNG option, as long as that's made clear to the player.

I dislike having a different default RNG for each difficulty. I think changing a core mechanic can be confusing for new players, and defaulting to no-RNG doesn't make it clear that full-RNG is the "intended" way to play. I'd rather default to full-RNG, but have no-RNG and biased-RNG as clearly visible options.

**Comment by doofus-01** [comment]
It would be good if EASY/NORMAL/HARD symbols had more default distinctions than gold and turns.  No RNG on EASY makes sense.

This might be out of scope, but regarding Dalas121 last point, can there be some consistent reminder of what RNG option is used?  Blue highlight frames instead of gold, or something like that.  Subtle and non-intrusive, but always there and easy to spot-check.

**Comment by stevecotton** [comment]
What would the UI be for this? Would changing the drop-down on the bottom-right of the campaign-select screen automatically change the drop-down to its left?

My general feeling is against this - EASY should be giving the player a good chance to recover from a mistake, and if it isn't then I think we'd get a better game by asking why it's not doing that, rather than automatically assuming that different RNG would be the answer.

**Comment by Pentarctagon** [comment]
> This might be out of scope, but regarding Dalas121 last point, can there be some consistent reminder of what RNG option is used? Blue highlight frames instead of gold, or something like that. Subtle and non-intrusive, but always there and easy to spot-check.

That would be possible.

> What would the UI be for this? Would changing the drop-down on the bottom-right of the campaign-select screen automatically change the drop-down to its left?

Yep. Also maybe with an "always use X RNG" option in the preferences.

> My general feeling is against this - EASY should be giving the player a good chance to recover from a mistake, and if it isn't then I think we'd get a better game by asking why it's not doing that, rather than automatically assuming that different RNG would be the answer.

I think the answer to that is fairly straight forward - new players who aren't used to the randomness at all, aren't used to it not being weighted in their favor (like many other strategy games tend to do, such as [Fire Emblem](https://fireemblemwiki.org/wiki/True_hit)), or they're just new to strategy games entirely. They then don't plan for the rare outcomes like a Mage missing all three of its attacks, which despite being unlikely isn't exceptionally uncommon either - it's 2.7%, or about 1 in 37 attacks with Mages will miss all three times. It's easier (for the player) to not have to worry about that yet and instead spend their effort learning the rest of the game's mechanics, and may also make it easier to understand in that seeing on the UI that they get a 70% damage penalty attacking that elvish archer on a forest is a much more direct presentation than a 30% chance to deal full damage, given how bad people's instincts about statistics tend to be.

**Comment by Discontinuum** [comment]
My couple of thoughts. New players may not be used to the Wesnoth randomness, but they won't ever be if they don't get exposed to it. This may be not that much of a problem for singleplayer, but for multiplayer (which supports only default RNG, as far as I know) I see a potential obstacle: a player who played only campaigns with no-RNG or biased RNG and never had an idea about the original Wesnoth RNG would have an extra shock when goes to multiplayer

Regarding default RNGs for difficulty levels, I have mixed feelings. I see the motivation to soften the RNG for newcomers, and the proposed change has its logic: players switches difficulties and RNG automatically switches to a more "difficult" version. But on the other hand, I think this may increase confusion from the UX prospective. Difficulty and type of RNG are not tied together (and they shouldn't ofc, a player should have a choice to play any difficulty with any RNG). And if switching difficulties will cause the RNG type to switch implicitly, this may present a UX problem I think (like "I changed difficulty but forgot to change RNG, damn, now I should restart again.") Maybe it's a subtle problem, still I want to highlight it at least.

**Comment by Pentarctagon** [comment]
Part of the reason to tie it to difficulty is also just for practicality - having all difficulties for all campaigns be balanced for all RNG types would be a massive amount of work.

**Comment by Dalas121** [comment]
I agree balancing should be done only for the default RNG setting at each difficulty, but I think that's equally compatible with any set of default RNG settings, whether easy/medium/hard=no-RNG/biased RNG/full-RNG, or easy/medium/hard=full-RNG.

**Comment by stevecotton** [comment]
> And if switching difficulties will cause the RNG type to switch implicitly, this may present a UX problem I think (like "I changed difficulty but forgot to change RNG, damn, now I should restart again.") Maybe it's a subtle problem, still I want to highlight it at least.

Alternative suggetions:

* highlight the RNG if it isn't the default for that difficulty, but only change the selected value if the player changes it manually.
* have a RNG setting called "default for difficulty", which causes the automatic mapping of difficulty to RNG. If the user changes it manually, don't alter it automatically.

> having all difficulties for all campaigns be balanced for all RNG types would be a massive amount of work.

Apologies if I'm underestimating the about of work that @nemaara did, but IIUC they were mainly balanced at one difficulty and then the other difficulties' stats were done by rule-of-thumb rather than testing.

**Comment by Pentarctagon** [comment]
My guess/assumption is that on harder difficulty the 100% chance of inflicting or being inflicted with poison or slow would have a much larger impact than it would on Easy.

**Comment by CelticMinstrel** [comment]
> My guess/assumption is that on harder difficulty the 100% chance of inflicting or being inflicted with poison or slow would have a much larger impact than it would on Easy, at least for campaigns which have those types of units.

I think giving a 100% chance of inflicting slow or poison with the biased RNG should be treated as a bug and fixed. Changing the RNG shouldn't suddenly make poison onstoppable. The biased RNG should just use the chance to hit as the chance to inflict the status effect.

That might be much easier said than done, though.

**Comment by nemaara** [comment]
> Apologies if I'm underestimating the amount of work that @nemaara did, but IIUC they were mainly balanced at one difficulty and then the other difficulties' stats were done by rule-of-thumb rather than testing.

I tended to test on two difficulties, one more rigorously than the other. Not for all campaigns, just for some that I had the energy to playtest multiple times.

As for the question at hand, I also tend toward not varying the default RNG based on difficulty, but haven't really tried less vs. more RNG enough to have much of an opinion either way. I will say I don't think I can rebalance everything very easily should we decide to change the RNG to the difficulty level though. And how would this work with the 6 levels of campaign difficulties from novice to nightmare? Something like UtBS already starts at challenging so would we still use the "noob" no RNG for them?

**Comment by Pentarctagon** [comment]
IMO if a campaign doesn't have an "easy" difficulty, then it would only ever default to Biased RNG or Default RNG.
